File: src/main/java/com/vc/medium/L969.scala
package com.vc.medium

object L969 {
  def pancakeSort(arr: Array[Int]): List[Int] = {
    if(arr.length == 1) return arr.toList

    def reverse(start: Int, end: Int): Unit = {
      var from = start
      var to = end
      while(from < to) {
        val tmp = arr(from)
        arr(from) = arr(to)
        arr(to) = tmp
        from += 1
        to -= 1
      }
    }

    def findMaxIdx(end: Int): Int = {
      var maxIdx = 0
      var max = arr(0)
      (1 to end).foreach(i => {
        if(arr(i) > max) {
          max = arr(i)
          maxIdx = i
        }
      })
      maxIdx
    }

    var end = arr.length - 1
    import scala.collection.mutable
    val res = new mutable.ArrayBuffer[Int]()
    while(end > 0) {
      val maxIdx = findMaxIdx(end)
      res += maxIdx + 1
      res += end + 1
      reverse(0, maxIdx)
      reverse(0, end)
      end -= 1
    }
    res.toList
  }
}

File: src/main/java/com/vc/medium/L969.java
package com.vc.medium;

import java.util.*;

class L969 {
    public List<Integer> pancakeSort(int[] A) {
        //Reverse the current max number to the first place and then reverse it to the correct place.
        int currentMax = A.length;
        List<Integer> res = new ArrayList<>();
        for(int i = 0; i < A.length; i++) {
            int index = findCurrentMax(A, currentMax);
            flip(A, index);
            flip(A, currentMax - 1);

            res.add(index + 1);
            res.add(currentMax--);
        }
        return res;
    }

    private int findCurrentMax(int[] A, int currentMax) {
        for(int i = 0; i < A.length; i++) {
            if(A[i] == currentMax) return i;
        }
        return -1;
    }

    private void flip(int[] A, int index) {
        int from = 0, to = index;
        while(from <= to) {
            int temp = A[from];
            A[from] = A[to];
            A[to] = temp;
            from++;
            to--;
        }
    }
}
